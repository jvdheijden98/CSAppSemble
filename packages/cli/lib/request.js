import { logger } from '@appsemble/node-utils';
import axios from 'axios';
import FormData from 'form-data';

/**
 * Make an HTTP request using Axios.
 *
 * @param {Request} req The Axios request.
 */
export default async function request(req) {
  logger.info(`Start ${req.method.toUpperCase()} ${req.url}`);
  logger.silly(`Request body: ${JSON.stringify(req.body)}`);
  const { data } = await axios(req);
  logger.info(`Success ${req.method.toUpperCase()} ${req.url}`);
  logger.verbose(`Response body: ${JSON.stringify(data)}`);
  return data;
}

/**
 * Make an HTTP POST request.
 *
 * @param {string} url The URL endpoint to make the POST request to.
 * @param data The data to send. If this is a `FormData` instance, headers will be handled
 *   automatically.
 * @param {Object} headers The HTTP request headers to send.
 * @returns The response body.
 * @throws If the request failed, or a bad HTTP status code was returned.
 */
export function post(url, data, headers = {}) {
  if (data instanceof FormData) {
    Object.assign(headers, data.getHeaders());
  }
  return request({ method: 'post', url, data, headers });
}

/**
 * Make an HTTP PATCH request.
 *
 * @param {string} url The URL endpoint to make the PATCH request to.
 * @param data The data to send. If this is a `FormData` instance, headers will be handled
 *   automatically.
 * @param {Object} headers The HTTP request headers to send.
 * @returns The response body.
 * @throws If the request failed, or a bad HTTP status code was returned.
 */
export function patch(url, data, headers = {}) {
  if (data instanceof FormData) {
    Object.assign(headers, data.getHeaders());
  }
  return request({ method: 'patch', url, data, headers });
}
