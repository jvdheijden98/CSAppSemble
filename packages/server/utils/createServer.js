#!/usr/bin/env node
import { loggerMiddleware } from '@appsemble/node-utils';
import faPkg from '@fortawesome/fontawesome-free/package.json';
import Koa from 'koa';
import compress from 'koa-compress';
import mount from 'koa-mount';
import koaQuerystring from 'koa-qs';
import range from 'koa-range';
import session from 'koa-session';
import serve from 'koa-static';
import koasBodyParser from 'koas-body-parser';
import koas from 'koas-core';
import koasOAuth2Server from 'koas-oauth2-server';
import koasOperations from 'koas-operations';
import koasParameters from 'koas-parameters';
import koasSerializer from 'koas-serializer';
import koasSpecHandler from 'koas-spec-handler';
import koasStatusCode from 'koas-status-code';
import koasSwaggerUI from 'koas-swagger-ui';
import path from 'path';
import raw from 'raw-body';

import api from '../api';
import * as operations from '../controllers';
import appMapper from '../middleware/appMapper';
import boom from '../middleware/boom';
import frontend from '../middleware/frontend';
import oauth2 from '../middleware/oauth2';
import tinyRouter from '../middleware/tinyRouter';
import { appRouter, fallbackRouter, studioRouter } from '../routes';
import bulmaHandler from '../routes/bulmaHandler';
import Mailer from './email/Mailer';
import oauth2Model from './oauth2Model';

export default async function createServer({
  app = new Koa(),
  argv = {},
  db,
  secret = 'appsemble',
  webpackConfigs,
}) {
  // eslint-disable-next-line no-param-reassign
  app.keys = [secret];
  app.use(loggerMiddleware());
  app.use(session(app));

  app.use(boom());
  app.use(range);
  Object.assign(app.context, { argv, db, mailer: new Mailer(argv) });

  if (argv.oauthGitlabKey || argv.oauthGoogleKey) {
    app.use(oauth2(argv));
  }

  koaQuerystring(app);

  if (process.env.NODE_ENV === 'production') {
    app.use(compress());
  }

  app.use(
    mount(
      `/fa/${faPkg.version}`,
      serve(path.dirname(require.resolve('@fortawesome/fontawesome-free/package.json'))),
    ),
  );

  app.use(
    tinyRouter([
      {
        route: /^\/bulma/,
        get: bulmaHandler,
      },
    ]),
  );

  app.use(
    await koas(api(), [
      koasSpecHandler(),
      koasSwaggerUI({ url: '/api-explorer' }),
      koasOAuth2Server(oauth2Model({ db, secret })),
      koasParameters(),
      koasBodyParser({
        '*/*': (body, mediaTypeObject, ctx) =>
          raw(body, {
            length: ctx.request.length,
          }),
      }),
      koasSerializer(),
      koasStatusCode(),
      koasOperations({ operations }),
    ]),
  );

  if (process.env.NODE_ENV !== 'test') {
    app.use(await frontend(webpackConfigs));
  }

  app.use(appMapper(studioRouter, appRouter, fallbackRouter));

  return app.callback();
}
