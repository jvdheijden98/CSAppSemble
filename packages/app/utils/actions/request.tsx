import {
  HTTPMethodsUpper,
  RequestAction,
  RequestLikeAction,
  RequestLikeActionDefinition,
  RequestLikeActionTypes,
} from '@appsemble/types';
import { compileFilters, MapperFunction, validate } from '@appsemble/utils';
import axios, { AxiosRequestConfig } from 'axios';

import { MakeActionParameters } from '../../types';
import uploadBlobs from '../uploadBlobs';

export function requestLikeAction<T extends RequestLikeActionTypes>({
  definition: { blobs = {}, method = 'GET', schema, query, url, serialize },
  onSuccess,
  onError,
}: MakeActionParameters<RequestLikeActionDefinition<T>>): RequestLikeAction<'request'> {
  const regex = /{(.+?)}/g;
  const urlMatch = url.match(regex);
  const mappers =
    urlMatch &&
    urlMatch
      .map(match => match.substring(1, match.length - 1))
      .reduce<Record<string, MapperFunction>>((acc, filter) => {
        acc[filter] = compileFilters(filter);
        return acc;
      }, {});

  return {
    type: 'request',
    async dispatch(data) {
      const methodUpper = method.toUpperCase() as HTTPMethodsUpper;
      const req: AxiosRequestConfig = {
        method: methodUpper,
        url: url.replace(regex, (_match, filter) => mappers[filter](data)),
        params: methodUpper === 'GET' ? { ...query, ...data } : query,
      };

      if (methodUpper === 'PUT' || methodUpper === 'POST' || methodUpper === 'PATCH') {
        let body;

        if (serialize && serialize === 'formdata') {
          const formData = new FormData();

          const processFormData = (key: string, value: Blob | any[] | object | string): void => {
            if (value instanceof Blob) {
              formData.append(key, value);
            } else if (Array.isArray(value)) {
              value.forEach(item => {
                // Recursively iterate over values
                processFormData(key, item);
              });
            } else if (value instanceof Object) {
              formData.append(key, JSON.stringify(value));
            } else {
              // Primitives
              formData.append(key, value);
            }
          };

          Object.entries(data as object).forEach(([key, value]) => {
            processFormData(key, value);
          });

          body = formData;
        } else {
          switch (blobs.type) {
            case 'upload': {
              body = await uploadBlobs(data, blobs);
              break;
            }
            default:
              body = data;
          }
        }

        if (schema) {
          await validate(schema, body);
        }

        req.data = body;
      }

      try {
        const response = await axios(req);

        if (onSuccess) {
          return onSuccess.dispatch(response.data);
        }

        return response.data;
      } catch (exception) {
        if (onError) {
          return onError.dispatch(exception);
        }

        throw exception;
      }
    },
    method,
    url,
  };
}

export default function request(
  args: MakeActionParameters<RequestLikeActionDefinition<'request'>>,
): RequestAction {
  return requestLikeAction(args);
}
