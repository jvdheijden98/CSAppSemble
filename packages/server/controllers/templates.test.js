import jwt from 'jsonwebtoken';
import lolex from 'lolex';
import request from 'supertest';

import createServer from '../utils/createServer';
import testSchema from '../utils/test/testSchema';
import testToken from '../utils/test/testToken';
import truncate from '../utils/test/truncate';

describe('Template API', () => {
  let db;
  let server;
  let token;
  let organizationId;
  let App;
  let Resource;
  let templates;

  let clock;

  beforeAll(async () => {
    db = await testSchema('templates');

    server = await createServer({ db });
    ({ App, Resource } = db.models);
  }, 10e3);

  beforeEach(async () => {
    await truncate(db);
    token = await testToken(server, db, 'apps:read apps:write');
    organizationId = jwt.decode(token.substring(7)).user.organizations[0].id;
    clock = lolex.install();

    const template = {
      path: 'test-template',
      template: true,
      vapidPublicKey: 'a',
      vapidPrivateKey: 'b',
      OrganizationId: organizationId,
      definition: {
        name: 'Test Template',
        description: 'Description',
        pages: [],
      },
    };

    const t1 = await App.create(template, { raw: true });
    const t2 = await App.create(
      {
        ...template,
        path: 'test-template-2',
        definition: { ...template.definition, name: 'Test App 2' },
        resources: {
          test: { schema: { type: 'object', properties: { name: { type: 'string' } } } },
        },
      },
      { raw: true },
    );
    await Resource.create({ type: 'test', data: { name: 'foo' }, AppId: t2.id });

    templates = [t1, t2];
  });

  afterEach(() => {
    clock.uninstall();
  });

  afterAll(async () => {
    await db.close();
  });

  it('should return a list of available templates', async () => {
    const { body: result } = await request(server).get('/api/templates');

    expect(result).toHaveLength(2);
    expect(result).toStrictEqual([
      {
        id: templates[0].id,
        name: templates[0].definition.name,
        description: templates[0].definition.description,
        resources: false,
      },
      {
        id: templates[1].id,
        name: templates[1].definition.name,
        description: templates[1].definition.description,
        resources: true,
      },
    ]);
  });

  it('should create a new app using a template', async () => {
    const { body: result } = await request(server)
      .post('/api/templates')
      .set('Authorization', token)
      .send({
        templateId: templates[0].id,
        name: 'Test app',
        description: 'This is a test app',
        organizationId,
      });

    expect(result).toMatchSnapshot();
  });

  it('should create a new app with example resources', async () => {
    const template = templates[1];
    const { body: result } = await request(server)
      .post('/api/templates')
      .set('Authorization', token)
      .send({
        templateId: template.id,
        name: 'Test app',
        description: 'This is a test app',
        organizationId,
        resources: true,
      });

    const { id } = result;
    const resources = await Resource.findAll({ where: { AppId: id, type: 'test' } }, { raw: true });

    expect(resources.map(r => r.data)).toStrictEqual([{ name: 'foo' }]);
  });

  it('should append a number when creating a new app using a template with a duplicate name', async () => {
    await request(server)
      .post('/api/templates')
      .set('Authorization', token)
      .send({
        templateId: templates[0].id,
        name: 'Test app',
        description: 'This is a test app',
        organizationId,
      });

    const {
      status,
      body: { path },
    } = await request(server)
      .post('/api/templates')
      .set('Authorization', token)
      .send({
        templateId: templates[0].id,
        name: 'Test app',
        description: 'This is also a test app',
        organizationId,
      });

    expect(status).toStrictEqual(201);
    expect(path).toStrictEqual('test-app-2');
  });

  it('should fall back to append random bytes to the end of the app path after 10 attempts', async () => {
    await Promise.all(
      [...new Array(11)].map((_, index) =>
        App.create(
          {
            path: index + 1 === 1 ? 'test-app' : `test-app-${index + 1}`,
            definition: { name: 'Test App', defaultPage: 'Test Page' },
            vapidPublicKey: 'a',
            vapidPrivateKey: 'b',
            OrganizationId: organizationId,
          },
          { raw: true },
        ),
      ),
    );

    const {
      status,
      body: { path },
    } = await request(server)
      .post('/api/templates')
      .set('Authorization', token)
      .send({
        templateId: templates[0].id,
        name: 'Test app',
        description: 'This is a test app',
        organizationId,
      });

    expect(status).toStrictEqual(201);
    expect(path).toMatch(/test-app-(\w){10}/);
  });
});
