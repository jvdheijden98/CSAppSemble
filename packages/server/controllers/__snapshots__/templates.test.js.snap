// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template API should create a new app using a template 1`] = `
Object {
  "$created": "1970-01-01T00:00:00.000Z",
  "$updated": "1970-01-01T00:00:00.000Z",
  "OrganizationId": "testorganization",
  "definition": Object {
    "description": "This is a test app",
    "name": "Test app",
    "pages": Array [],
  },
  "domain": null,
  "iconUrl": "/api/apps/5/icon",
  "id": 5,
  "path": "test-app",
  "private": false,
  "yaml": "name: Test app
description: This is a test app
pages: []
",
}
`;
